src/MongoDB.AspNetCore.OData/MongoEnableQueryAttribute.cs

```diff
@@ -0,0 +1,141 @@
+using MongoDB.Driver;
+using MongoDB.Driver.Linq;
+
+namespace MongoDB.AspNetCore.OData;
+
+public sealed class MongoEnableQueryAttribute : EnableQueryAttribute
+{
+    private static readonly MethodInfo s_applyQueryInternalMethodInfo = typeof(MongoEnableQueryAttribute).GetMethod(
+        nameof(ApplyQueryInternal),
+        BindingFlags.Static | BindingFlags.NonPublic);
+    private static readonly MethodInfo s_applyProjectionMethodInfo = typeof(MongoEnableQueryAttribute).GetMethod(
+        nameof(ApplyProjection),
+        BindingFlags.Static | BindingFlags.NonPublic);
+
+    public MongoEnableQueryAttribute()
+    {
+        AllowedQueryOptions =
+            AllowedQueryOptions.Filter |
+            AllowedQueryOptions.Expand |
+            AllowedQueryOptions.Select |
+            AllowedQueryOptions.OrderBy |
+            AllowedQueryOptions.Top |
+            AllowedQueryOptions.Skip |
+            AllowedQueryOptions.Count |
+            AllowedQueryOptions.Compute;
+    }
+
+    public override void ValidateQuery(HttpRequest request, ODataQueryOptions queryOptions)
+    {
+        base.ValidateQuery(request, queryOptions);
+
+        // Store this into the http context in case we will need to enforce projection
+        request.HttpContext.Items[nameof(ODataQueryOptions)] = queryOptions;
+    }
+
+    public override void OnActionExecuted(ActionExecutedContext actionExecutedContext)
+    {
+        base.OnActionExecuted(actionExecutedContext);
+
+        var response = actionExecutedContext.HttpContext.Response;
+        // Check whether the response is set and successful
+        if (response != null && response.IsSuccessStatusCode() && actionExecutedContext.Result is ObjectResult responseContent)
+        {
+            // If response is IMongoQueryable it means default implementation bypasses all transformations,
+            // we want force apply projection to limit MQL request to struct properties only
+            if (responseContent.Value is IMongoQueryable mongoQueryable)
+            {
+                var queryOptions = (ODataQueryOptions)response.HttpContext.Items[nameof(ODataQueryOptions)];
+                var applyProjectionMethod = s_applyProjectionMethodInfo.MakeGenericMethod(mongoQueryable.ElementType);
+                responseContent.Value = applyProjectionMethod.Invoke(null, new object[] { mongoQueryable, queryOptions });
+            }
+        }
+    }
+
+    public override IQueryable ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)
+    {
+        if (queryable is IMongoQueryable)
+        {
+            var applyImplMethodInfo = s_applyQueryInternalMethodInfo.MakeGenericMethod(queryable.ElementType);
+            return (IQueryable)applyImplMethodInfo.Invoke(null, new object[] { queryable, queryOptions });
+        }
+
+        return base.ApplyQuery(queryable, queryOptions);
+    }
+
+    private static IQueryable ApplyQueryInternal<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
```

Comment:
Consider `IMongoQueryable<T>` for same reasons as in `ApplyProjection`

Reply:
Done